pkg/daytona/unmarshal.go

```diff
@@ -0,0 +1,251 @@
+const (
+	vaultTagPath = "vault_path"
+	vaultTagKey  = "vault_key"
+
+	vaultTagPathKey = "vault_path_key"
+
+	defaultVaultTagField = "value"
+)
+
+var ErrValueInput = errors.New("the provided value must be a struct pointer")
+
+type unmarshalSecretsConfig struct {
+	apex string
+}
+
+// An UnmarshalSecretsOption is an option for the UnmarshalSecrets function
+type UnmarshalSecretsOption interface {
+	apply(c *unmarshalSecretsConfig)
+}
+
+// WithApex returns an UnmarshalSecretsOption sets the apex value
+// for which the secrets unmarshaler should use
+func WithApex(apex string) UnmarshalSecretsOption {
+	return withApex(apex)
+}
+
+type withApex string
+
+func (w withApex) apply(c *unmarshalSecretsConfig) {
+	c.apex = strings.TrimSuffix(string(w), "/")
+}
+
+// UnmarshalSecrets traverses the value v recursively looking for tagged fields that
+// can be populated with secret data using the provided client and optional configured apex.
+// If the apex is configured, the tag vault_path_key is appended to the apex to construct
+// the final secret path. If the tag vault_path is provided, the apex is ignored.
+// A default key of 'default' is used on each path, it can be overridden using the vault_key tag.
+//
+//  Apex of 'secret/application' provided, combined to form 'secret/application/db_password'
+//  Field string `vault_path_key:"db_password"`
+//
+//  Apex of 'secret/application' provided, with key override
+//  Field string `vault_path_key:"db_password" vault_key:"password"`
+//
+//  vault_path represents a full secret path to fetch
+//  Field string `vault_path:"secret/application/db_password"`
+//  Field string `vault_path:"secret/application/db_password" vault_key:"password"` // key override
+func UnmarshalSecrets(client *api.Client, v interface{}, opts ...UnmarshalSecretsOption) error {
+	config := unmarshalSecretsConfig{}
+
+	for _, opt := range opts {
+		opt.apply(&config)
+	}
+
+	val := reflect.ValueOf(v)
+	if val.Kind() != reflect.Ptr {
+		return ErrValueInput
+	}
+
+	val = val.Elem()
+	if val.Kind() != reflect.Struct {
+		return ErrValueInput
+	}
+
+	for i := 0; i < val.NumField(); i++ {
+		fName := val.Type().Field(i).Name
+		f := val.Field(i)
+		if f.Kind() == reflect.Ptr {
+			if f.IsNil() {
+				f.Set(reflect.New(f.Type().Elem()))
+			}
+			f = f.Elem()
+		}
+
+		path, valueIndex := parsePath(config.apex, val.Type().Field(i).Tag)
+		switch f.Kind() {
+		case reflect.Struct:
+			err := UnmarshalSecrets(client, f.Addr().Interface(), opts...)
+			if err != nil {
+				return err
+			}
+		case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+			if path == "" {
+				continue
+			}
+
+			var iv int64
+
+			v, err := fetchValue(client, path, valueIndex)
+			if err != nil {
+				return err
+			}
+
+			if f.Kind() == reflect.Int64 && f.Type().String() == "time.Duration" {
+				var d time.Duration
+				dur, ok := v.(string)
+				if ok {
+					d, err = time.ParseDuration(dur)
+					if err != nil {
+						return err
```

Comment:
So I always hate it when I am using a library like this and it returns an error, I fix that error, then it returns another error. Should we consider using something like https://pkg.go.dev/go.uber.org/multierr to get all the errors from the config and return them all at once?